crates/oxide/src/glob.rs

```diff
@@ -1,11 +1,117 @@
+/// base of the PublicSourceEntry.
+///
+/// ```diff
+/// - { base: '/', pattern: 'src/**/*.html'}
+/// + { base: '/src', pattern: '**/*.html'}
+/// ```
+///
+/// A file stays in the `pattern` part, because the `base` should only be a directory.
+///
+/// ```diff
+/// - { base: '/', pattern: 'src/examples/index.html'}
+/// + { base: '/src/examples', pattern: 'index.html'}
+/// ```
+///
+/// A folder will be moved to the `base` part, and the `pattern` will be set to `**/*`.
+///
+/// ```diff
+/// - { base: '/', pattern: 'src/examples'}
+/// + { base: '/src/examples', pattern: '**/*'}
+/// ```
+///
+/// In addition, we will canonicalize the base path so we always work with the correctly resolved
+/// path.
+pub fn optimize_public_source_entry(source: &mut PublicSourceEntry) {
+    // Resolve base path immediately
+    let Ok(base) = dunce::canonicalize(&source.base) else {
+        event!(Level::ERROR, "Failed to resolve base: {:?}", source.base);
+        return;
+    };
+    source.base = base.to_string_lossy().to_string();
+
+    // No dynamic part, figure out if we are dealing with a file or a directory.
+    if !source.pattern.contains('*') {
+        let combined_path = if source.pattern.starts_with("/") {
+            PathBuf::from(&source.pattern)
+        } else {
+            PathBuf::from(&source.base).join(&source.pattern)
+        };
+
+        match dunce::canonicalize(combined_path) {
+            Ok(resolved_path) if resolved_path.is_dir() => {
+                source.base = resolved_path.to_string_lossy().to_string();
+                source.pattern = "**/*".to_owned();
+            }
+            Ok(resolved_path) if resolved_path.is_file() => {
+                source.base = resolved_path
+                    .parent()
+                    .unwrap()
+                    .to_string_lossy()
+                    .to_string();
+                // Ensure leading slash, otherwise it will match against all files in all folders/
+                source.pattern = format!(
+                    "/{}",
+                    resolved_path
+                        .file_name()
+                        .unwrap()
+                        .to_string_lossy()
+                        .to_string()
+                );
+            }
+            _ => {}
+        }
+        return;
+    }
+
+    // Contains dynamic part
+    let (static_part, dynamic_part) = split_pattern(&source.pattern);
+
+    let base: PathBuf = source.base.clone().into();
+    let base = match static_part {
+        Some(static_part) => base.join(static_part),
+        None => base,
+    };
+
+    // TODO: If the base does not exist on disk, try removing the last slash and try again.
+    let base = match dunce::canonicalize(&base) {
+        Ok(base) => base,
+        Err(err) => {
+            event!(tracing::Level::ERROR, "Failed to resolve glob: {:?}", err);
+            return;
+        }
+    };
```

Comment:
nit: this runs canonicalize twice on what is likely the same string in the case of the pattern being `**/*` or `*.js` or whatever

Maybe something we can improve? Not a big deal tho. It's gonna hit the fs cache in the OS I'd imagine.